core/translate/planner.rs

```diff
@@ -0,0 +1,339 @@
+use super::node::{Aggregate, Direction, QueryPlanNode};
+use crate::{
+    function::Func,
+    schema::{BTreeTable, Schema},
+    util::normalize_ident,
+    Result,
+};
+use sqlite3_parser::ast::{self, FromClause, JoinType, ResultColumn};
+use std::fmt::Display;
+use std::rc::Rc;
+
+pub struct NodeIdCounter {
+    id: usize,
+}
+
+impl NodeIdCounter {
+    pub fn new() -> Self {
+        Self { id: 0 }
+    }
+    pub fn get_next_id(&mut self) -> usize {
+        let id = self.id;
+        self.id += 1;
+        id
+    }
+}
+
+pub struct SelectPlan {
```

Comment:
I would perhaps call this `Plan` and move it to `node.rs` (which I think should probably be renamed to `plan.rs`).